### PR TITLE
feat: add "before" actions

### LINKED ISSUE #68: Add support for "before" hooks
For some flows it can be useful to run action before command. For example if to dump data or database before running command. I think for this we will need to add actions in settings and add execution before this lines [beforeReset](https://github.com/graphile/migrate/blob/master/src/commands/reset.ts#L19), [beforeAllMigrations](https://github.com/graphile/migrate/blob/master/src/commands/migrate.ts#L36) and [beforeCurrent](https://github.com/graphile/migrate/blob/master/src/commands/watch.ts#L41).

For `beforeCurrent` there is some question when should it run? As there is a check to see if file have been changed and if we want to run `beforeCurrent` only when file is changed then we will ned to run it in transaction [here](https://github.com/graphile/migrate/blob/master/src/commands/watch.ts#L83)?

Is there something I'm missing?

### PR DESCRIPTION
**Pull request by hos**
Feature implementation for #68

Fixes #68

### CODE REVIEW

File: src/commands/migrate.ts [hunk 1]
```diff
@@ -33,6 +33,13 @@ export async function _migrate(
           parsedSettings,
           lastMigration,
         );
+        if (remainingMigrations.length > 0) {
```
Comment (benjie) [inline src/commands/migrate.ts:36]:
```suggestion
        if (remainingMigrations.length > 0 || forceActions) {
```

File: src/commands/watch.ts [hunk 2]
```diff
@@ -80,6 +80,14 @@ export function _makeCurrentMigrationRunner(
 
             // 4: if different
             if (!migrationsAreEquivalent) {
+              if (parsedSettings.beforeCurrent) {
+                await executeActions(
+                  parsedSettings,
+                  shadow,
+                  parsedSettings.beforeCurrent,
+                );
+              }
```
Comment (benjie) [inline src/commands/watch.ts:89]: I think this is the most sensible place for this currently.

Comment (hos) [reply src/commands/watch.ts:89]: Yes, and it is was what I needed, but added the others for completeness. Even it is yet v0, I don't think we want to add them and remove later. So let me know if you want, I will remove the other hooks.

File: README.md [hunk 3]
```diff
@@ -424,14 +424,23 @@ Configuration goes in `.gmrc`, which is a JSON file with the following keys:
   values) — graphile-migrate does not perform any escaping for you. The special
   value `!ENV` will tell graphile-migrate to load the setting from the
   environment variable with the same name.
+- `beforeReset` — optional list of actions to execute before creating the
+  database, can be used to run save before database will be dropped with reset.
```
Comment (benjie) [inline README.md:428]:
```suggestion
- `beforeReset` — optional list of actions to execute before deleting and
  recreating the database.
```

File: README.md [hunk 4]
```diff
@@ -424,14 +424,23 @@ Configuration goes in `.gmrc`, which is a JSON file with the following keys:
...
 - `afterReset` — optional list of actions to execute after the database has been
   created but before the migrations run, useful to set default permissions,
   install extensions or install external schemas like `graphile-worker` that
   your migrations may depend on. See "Actions" below.
+- `beforeAllMigrations` — optional list of actions to execute before the first
+  migration will ran, can be used to save some information about database state
+  before running migrations, for example collect stats.
```
Comment (benjie) [inline README.md:435]:
```suggestion
- `beforeAllMigrations` — optional list of actions to execute before any
  pending migrations are executed.
```

File: README.md [hunk 5]
```diff
@@ -424,14 +424,23 @@ Configuration goes in `.gmrc`, which is a JSON file with the following keys:
...
 - `afterAllMigrations` — optional list of actions to execute after all the
   migrations have ran, useful for performing a tasks like dumping the database
   or regenerating dependent data (GraphQL schema, type definitions, etc). See
   "Actions" below.
+- `beforeCurrent` — optional list of actions to execute before `current.sql` is
+  executed. Can be used to save data before running `current.sql` and load again
+  with `afterCurrent`. This way you don't have to add test data after each
+  `current.sql` is executed.
```
Comment (benjie) [inline README.md:443]:
```suggestion
- `beforeCurrent` — optional list of actions to execute before `current.sql` is
  executed.
```

Comment (benjie) [reply README.md:443]: Particularly here if we put this in the description people will ask how to do it and I'm not sure I can come up with a generic way for this to work in all cases right now. I've therefore removed all the usage suggestions.

File: src/commands/migrate.ts [hunk 6]
```diff
@@ -33,6 +33,13 @@ export async function _migrate(
           parsedSettings,
           lastMigration,
         );
+        if (remainingMigrations.length > 0 || forceActions) {
```
Comment (benjie) [inline src/commands/migrate.ts:36]:
```suggestion
        const shouldExecuteActions = remainingMigrations.length > 0 || forceActions;
        if (shouldExecuteActions) {
```

Comment (benjie) [reply src/commands/migrate.ts:36]: Please update `if (remainingMigrations.length > 0 || forceActions)` for `afterAllMigrations` below to be `if (shouldExecuteActions)`. Better to have this shared logic in one place.

File: src/commands/watch.ts (hunk 2 above)
Comment (benjie) [inline src/commands/watch.ts:89]:
```suggestion
              await executeActions(
                parsedSettings,
                shadow,
                parsedSettings.beforeCurrent,
              );
```

I don't think this `if` is necessary - `executeActions` handles that for us.

Comment (benjie) [reply src/commands/watch.ts:89]: Also it will never be falsy, as can be seen from the types.

**Comment by benjie** [comment]
Well; just this one change now because I could auto-apply the others :)

**Comment by benjie** [comment]
(Also please run `yarn lint:fix` in case I broke formatting.)

**Pull request merged**